src/topp/DatabaseSuitability.cpp

```diff
@@ -0,0 +1,367 @@
+    OPENMS_LOG_INFO << count_ms2_lvl << " ms2 spectra." << endl;
+    OPENMS_LOG_INFO << "MS2 ID rate: " << id_rate << endl << endl;
+
+    if (!out.empty())
+    {
+      OPENMS_LOG_INFO << "Writting output to: " << out << endl << endl;
+
+      std::ofstream os(out);
+      os.precision(writtenDigits(double()));
+      os << "#top_db_hits\t" << count_db << endl;
+      os << "#top_novo_hits\t" << count_novo << endl;
+      os << "db_suitability\t" << quality << endl;
+      os << "#total_novo_seqs\t" << count_novo_seq << endl;
+      os << "#unique_novo_seqs\t" << unique_novo.size() << endl;
+      os << "#ms2_spectra" << count_ms2_lvl << endl;
+      os << "ms2_id_rate\t" << id_rate << endl;
+      os.close();
+    }
+
+    return EXECUTION_OK;
+
+  }
+
+private:
+
+  double getDecoyDiff_(const PeptideIdentification& pep_id)
+  {
+    double diff = -1;
+
+    // get the score of the first two decoy hits
+    double decoy_1 = -1;
+    double decoy_2 = -1;
+    UInt curr_hit = 1;
+
+    for (const auto& hit : pep_id.getHits())
+    {
+      if (curr_hit > 10) break;
+      ++curr_hit;
+
+      if (!hit.metaValueExists("target_decoy"))
+      {
+        throw(Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No target/decoy information found! Make sure 'PeptideIndexer' is run before hand."));
+      }
+
+      if (!hit.metaValueExists("MS:1002252"))
+      {
+        throw(Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No cross correlation score found at peptide hit. Only Comet search engine is supported right now."));
+      }
+
+      if (decoy_1 == -1 && hit.getMetaValue("target_decoy") == "decoy")
+      {
+        decoy_1 = hit.getMetaValue("MS:1002252");
+        continue;
+      }
+      if (decoy_1 > 0 && hit.getMetaValue("target_decoy") == "decoy")
+      {
+        decoy_2 = hit.getMetaValue("MS:1002252");
+        break;
+      }
+    }
+
+    if (decoy_2 > 0) // if there are two decoy hits
+    {
+      diff = abs(decoy_1 - decoy_2) / pep_id.getMZ(); // normalized by mw
+    }
+
+    // if there aren't two decoy hits -1 is returned
+    return diff;
+  }
+
+  double getDecoyCutOff_(const vector<PeptideIdentification>& pep_ids, double novor_fract)
+  {
+    double cut_off = -1;
+
+    // get all decoy diffs of peptide ids with at least two decoy hits
+    vector<double> diffs;
+    for (const auto& pep_id : pep_ids)
+    {
+      double diff = getDecoyDiff_(pep_id);
+      if (diff > 0)
+      {
+        diffs.push_back(diff);
+      }
+    }
+
+    if (double(diffs.size()) / pep_ids.size() < 0.2)
+    {
+      throw(Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "Under 20 % of peptide identifications have two decoy hits. This is not enough for re-ranking. Use the 'force_no_re_rank' flag to still compute a suitability score."));
+    }
+
+    // sort the diffs decreasing
+    std::sort(diffs.begin(), diffs.end(), std::greater<double>());
```

Comment:
partial_sort is enough?

Reply:
simply sort the first `novor_fract`*N of N positions (or `(1-novor_fract)*N in reverse order to save on sorting effort). And then grab the diff at that very position.
No need for the two loops below!?